src/host/_stream.cpp

```diff
@@ -126,565 +165,286 @@ void AdjustCursorPosition(SCREEN_INFORMATION& screenInfo, _In_ til::point coordC
-                                        gsl::narrow_cast<size_t>(coordScreenBufferSize.width) - sOriginalXPosition,
-                                        sOriginalXPosition,
-                                        dwFlags & WC_PRINTABLE_CONTROL_CHARS))
-                {
-                    CursorPosition.x = coordScreenBufferSize.width - 1;
-                    CursorPosition.y = cursor.GetPosition().y - 1;
-
-                    // since you just backspaced yourself back up into the previous row, unset the wrap flag
-                    // on the prev row if it was set
-                    textBuffer.GetRowByOffset(CursorPosition.y).SetWrapForced(false);
-
-                    AdjustCursorPosition(screenInfo, CursorPosition, dwFlags & WC_KEEP_CURSOR_VISIBLE, psScrollY);
-                }
+                const auto pos = cursor.GetPosition();
+                const auto tabCount = gsl::narrow_cast<size_t>(NUMBER_OF_SPACES_IN_TAB(pos.x));
+                numSpaces += _writeCharsLegacyUnprocessed(screenInfo, dwFlags, psScrollY, { &tabSpaces[0], tabCount });
+                continue;
             }
-            // Notify accessibility to read the backspaced character.
-            // See GH:12735, MSFT:31748387
-            if (screenInfo.HasAccessibilityEventing())
+            case UNICODE_LINEFEED:
             {
-                if (auto pConsoleWindow = ServiceLocator::LocateConsoleWindow())
+                auto pos = cursor.GetPosition();
+                if (WI_IsFlagClear(screenInfo.OutputMode, DISABLE_NEWLINE_AUTO_RETURN))
                 {
-                    LOG_IF_FAILED(pConsoleWindow->SignalUia(UIA_Text_TextChangedEventId));
+                    pos.x = 0;
                 }
-            }
-            break;
-        }
-        case UNICODE_TAB:
-        {
-            const auto TabSize = NUMBER_OF_SPACES_IN_TAB(cursor.GetPosition().x);
-            CursorPosition.x = cursor.GetPosition().x + TabSize;
-
-            // move cursor forward to next tab stop.  fill space with blanks.
-            // we get here when the tab extends beyond the right edge of the
-            // window.  if the tab goes wraps the line, set the cursor to the first
-            // position in the next line.
-            pwchBuffer++;
-
-            TempNumSpaces += TabSize;
-            size_t NumChars = 0;
-            if (CursorPosition.x >= coordScreenBufferSize.width)
-            {
-                NumChars = gsl::narrow<size_t>(coordScreenBufferSize.width - cursor.GetPosition().x);
-                CursorPosition.x = 0;
-                CursorPosition.y = cursor.GetPosition().y + 1;
 
-                // since you just tabbed yourself past the end of the row, set the wrap
-                textBuffer.GetRowByOffset(cursor.GetPosition().y).SetWrapForced(true);
+                textBuffer.GetRowByOffset(pos.y).SetWrapForced(false);
+                pos.y = pos.y + 1;

```

Comment:
we did an adjustment earlier for delayed cursor position. Does this accidentally make us commit TWO newlines when we're in delayed and emit a LF?

Reply:
I think I have a test build of terminal off a pre-merge branch, I **might** be able to try to repro this...

Reply:
I can't repro the thing I was worried about, but now I don't understand why ;D

Reply:
Code that I thought would break:

```cpp
#include <windows.h>
int main() {
	CONSOLE_SCREEN_BUFFER_INFO csbi;
	HANDLE h = GetStdHandle(STD_OUTPUT_HANDLE);
	GetConsoleScreenBufferInfo(h, &csbi);
	DWORD m;
	GetConsoleMode(h, &m);
	SetConsoleMode(h, ENABLE_PROCESSED_OUTPUT|DISABLE_NEWLINE_AUTO_RETURN);
	WriteConsoleW(h, L"\r\n", 2, NULL, NULL);
	for(int i = 0; i < csbi.dwSize.X-1; ++i) {
		WriteConsoleW(h, L".", 1, NULL, NULL);
	}
	WriteConsoleW(h, L"X", 1, NULL, NULL);
	WriteConsoleW(h, L"\n", 1, NULL, NULL);
	WriteConsoleW(h, L"\rGAP?\r\n", 7, NULL, NULL);
	SetConsoleMode(h, m);
	return 0;
}
```

what I was worried would happen:

```
.....X
<- this blank line should not be here
GAP
```

It didn't happen. But why?

Reply:
I think you're missing the `ENABLE_WRAP_AT_EOL_OUTPUT` flag. Although this case doesn't work correctly over conpty anyway, because there isn't a VT equivalent to the Windows console style of wrapping (at least as far as I know). I think we'd have to manually force the cursor position to move to the next line to make it work.

Reply:
Ah - I explicitly left it off because I was trying to get "delayed" behavior. Did I get the sense of that one backwards?

Reply:
I just realised now that it's weirder than I initially thought. Over conpty the blank line ends up _below_ the "GAP" text, i.e. something like this:

```
.....X
GAP
<- extra blank line here
```

Although I suppose that does make sense. The "GAP" is rendered in the wrong place, but conhost still knows where the cursor is meant to be, so the next time it actually needs to move the cursor (e.g. to draw the prompt), it's going to set it to the correct position.

Reply:
Oh I see. It's a combo flag. Ugh, I should know this stuff. Thanks!

Reply:
Okay, it **does** introduce a gap! But it also does so on conhost 22621. It likely doesn't matter. It's not a new bug, it's an existing caveat.

Reply:
As far as I can tell this works identical to the current WCL though, right? 
https://github.com/microsoft/terminal/blob/358e10b17fbf3737936139edc1d2d513450b9cb7/src/host/_stream.cpp#L157-L178

It first does the same delay wrap branch and then enters the primary loop which may assemble LFs.

Reply:
Just to be clear, when using the legacy console APIs you're _meant_ to get a gap. The delayed wrap thing is a VT-specific feature that doesn't apply here. We should possibly even be dropping these delayed wrap checks from `WriteCharsLegacy` because that flag should never be set in normal legacy usage - it's only likely to occur if you're mixing and matching VT output with legacy output, in which case there's no expected "correct" behavior.

The bug is that we're not getting the gap in the right place in Windows Terminal, at least as far as I can see.